{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import itertools\n",
    "import numpy as np\n",
    "from collections import Counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "def ReverseComplement(Pattern):\n",
    "    \"\"\"\n",
    "    Computes the reverse complement of a DNA Patternuence\n",
    "    ------\n",
    "    Inputs: \n",
    "    ------\n",
    "    Pattern(string): A nucleotide Patternuence for which the reverse complement\n",
    "    strand is to be determined, direction is 5' to 3'\n",
    "    ------\n",
    "    Outputs:\n",
    "    -------\n",
    "    rev_compl(string): A nucleotide Patternuence representing the reverese\n",
    "    complement of the original strand.\n",
    "    \"\"\"\n",
    "    # Convert Patternuence into the upper case letters\n",
    "    Pattern = Pattern.upper()\n",
    "    rev_compl = []\n",
    "\n",
    "    for ind in range(len(Pattern)):\n",
    "\n",
    "        if Pattern[ind] == 'A':\n",
    "            rev_compl.append('T')\n",
    "\n",
    "        elif Pattern[ind] == 'T':\n",
    "            rev_compl.append('A')\n",
    "\n",
    "        elif Pattern[ind] == 'C':\n",
    "            rev_compl.append('G')\n",
    "\n",
    "        elif Pattern[ind] == 'G':\n",
    "            rev_compl.append('C')\n",
    "\n",
    "    rev_compl.reverse()\n",
    "    rev_compl_str = ''.join(map(str, rev_compl))\n",
    "\n",
    "    return rev_compl_str\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def PatternCount(Text, Pattern):\n",
    "    \"\"\"\n",
    "    Counts the number of occurences of string Pattern\n",
    "    in the template Text.\n",
    "    ------\n",
    "    Inputs: \n",
    "    ------\n",
    "    Text(string): A sequence in which the pattern should be searched for.\n",
    "    Pattern(string): A sequence that is being searched for in 'Text'.\n",
    "    ------\n",
    "    Outputs:\n",
    "    -------\n",
    "    count(string): Number of occurences of the Pattern in String.\n",
    "    \"\"\"\n",
    "\n",
    "    count = 0\n",
    "\n",
    "    lT = len(Text)\n",
    "    lP = len(Pattern)\n",
    "\n",
    "    for i in range(lT - lP + 1):\n",
    "        if Text[i:i+lP] == Pattern:\n",
    "            count = count+1\n",
    "\n",
    "    return count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "def PatternMatching(Pattern, Genome):\n",
    "    \"\"\"\n",
    "    Computes the reverse complement of a DNA Patternuence\n",
    "    ------\n",
    "    Inputs: \n",
    "    ------\n",
    "    - Pattern(string): A nucleotide sequence for which\n",
    "    the match within the genome should be determined\n",
    "\n",
    "    - Genome(string): A nucleotide sequence within which\n",
    "    the occurences of the patern should be searched for\n",
    "    ------\n",
    "    Outputs:\n",
    "    -------\n",
    "    match_ind(list): All starting positions where Pattern \n",
    "    appears as a substring of Genome\n",
    "    \"\"\"\n",
    "\n",
    "    # Convert Pattern and Genome into the upper case letters,\n",
    "    # just in case\n",
    "    Pattern = Pattern.upper()\n",
    "    Genome = Genome.upper()\n",
    "\n",
    "\n",
    "    # Define the starting index\n",
    "    start = 0\n",
    "    # Get length of genome and pattern for iteration\n",
    "    gen_len = len(Genome)\n",
    "    pat_len = len(Pattern)\n",
    "\n",
    "    # Define the empty list of matched indices\n",
    "    match_ind = []\n",
    "\n",
    "    # Iterate over the Genome string\n",
    "    for ind in range(gen_len-pat_len+1):\n",
    "\n",
    "        if Genome[ind:ind+pat_len] == Pattern:\n",
    "            match_ind.append(ind)\n",
    "\n",
    "\n",
    "    return match_ind"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "def FrequentWords(Text, k):\n",
    "    \"\"\"\n",
    "    Find all the most frequent k-mers in a string\n",
    "    ------\n",
    "    Inputs: \n",
    "    ------\n",
    "    Text(string): A sequence in which the k-mers should be searched for.\n",
    "    k(int): length of the k-mer\n",
    "    ------\n",
    "    Outputs:\n",
    "    -------\n",
    "    freq_kmers(string): Lengths of the most frequent k-mers.\n",
    "    \"\"\"\n",
    "\n",
    "    # Initialize the empty variables\n",
    "    freq_kmers = list()\n",
    "    counts = list()\n",
    "\n",
    "\n",
    "    # Iterate over the 'Text' to find the patterns of k-mer length\n",
    "    for i in range(len(Text) - k):\n",
    "        kmer = Text[i:i+k]\n",
    "        counts.append(PatternCount(Text, kmer))\n",
    "\n",
    "        # Get the longest pattern\n",
    "        maxCount = max(counts)\n",
    "\n",
    "    for j in range(len(Text) - k):\n",
    "        if counts[j] == maxCount:\n",
    "            freq_kmers.append(Text[j:j+k])\n",
    "\n",
    "    # Remove duplicates\n",
    "    freq_kmers = list(set(freq_kmers))\n",
    "\n",
    "    return freq_kmers\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "def SymbolToNumber(base):\n",
    "    \"\"\"\n",
    "    Transforms the base into respective integer:\n",
    "    A -> 0\n",
    "    C -> 1\n",
    "    G -> 2\n",
    "    T -> 3\n",
    "    -------\n",
    "    Inputs:\n",
    "    -------\n",
    "    base(string): A nucleotide that is last in the string of nucleotides to be transformed\n",
    "    -------\n",
    "    Outputs:\n",
    "    -------\n",
    "    A number corresponding to the given base.\n",
    "    \"\"\"\n",
    "    \n",
    "    # Define a simple dictionary of bases and select one usint get() method\n",
    "    # for Python dictionary\n",
    "    base_dict = {'A':0, 'C':1, 'G':2, 'T':3}\n",
    "    return base_dict.get(base)\n",
    "\n",
    "def PatternToNumber(Pattern):\n",
    "    \"\"\"\n",
    "    Computes the frequency array\n",
    "    ------\n",
    "    Inputs: \n",
    "    ------\n",
    "    Pattern(string): A nucleotide sequence for which the frequency array\n",
    "    is to be determined\n",
    "    ------\n",
    "    Outputs:\n",
    "    -------\n",
    "    Integer with the lexicographic position of the given k-mer.\n",
    "    \"\"\"\n",
    "    # If pattern contains no symbols return zero\n",
    "    if Pattern == \"\":\n",
    "        return 0\n",
    "    \n",
    "    # Get the terminal nucleotide\n",
    "    symbol = Pattern[-1]\n",
    "    # Get the previous pattern\n",
    "    prefix = Pattern[:-1]\n",
    "\n",
    "    return 4*PatternToNumber(prefix) + SymbolToNumber(symbol)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "def NumberToPattern(index, k):\n",
    "    \"\"\"\n",
    "    Returns the nucleotide sequence at position index,\n",
    "    given the length of the k-mer 'k'.\n",
    "    ------\n",
    "    Inputs: \n",
    "    ------\n",
    "    index(int): position of the nucl. seq. in the frequency array\n",
    "    k(int): length of the desired k-mer\n",
    "    ------\n",
    "    Outputs:\n",
    "    -------\n",
    "    seq(string): the nucleotide sequence of the lexicographically\n",
    "    ordered k-mers at the position index.\n",
    "    \"\"\"\n",
    "    # Initialize the empty variables\n",
    "    kmer_list = list(range(4**(k)))\n",
    "\n",
    "    # Make all the permutations of length of Pattern using the nucleotides\n",
    "    kmer_list = [''.join(map(str, i)) for i in itertools.product(['A','C','T','G'], repeat=k)]\n",
    "\n",
    "    # Sort the list lexicographically\n",
    "    kmer_list.sort()\n",
    "\n",
    "\n",
    "    # Get the nucleotide at the given position\n",
    "    # indexing from 0\n",
    "    seq = kmer_list[index]\n",
    "\n",
    "    return seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "def ComputingFrequencies(Text, k):\n",
    "    \"\"\"\n",
    "    Generates the frequency array by first initializing every\n",
    "    element in the frequency array to zero (4^k operations) and\n",
    "    then making a single pass down Text (~ len(Text)*k operations),\n",
    "    increasing the value of the frequency array counter\n",
    "    corresponding to Patter for each k-mer Pattern encountered.\n",
    "\n",
    "    k-mer beginning at position i of Text:= Text(i,k)\n",
    "    ------\n",
    "    Inputs: \n",
    "    ------\n",
    "    Text(string): A DNA string for which the frequency array\n",
    "    is computed.\n",
    "    k(int): k-mer length\n",
    "    ------\n",
    "    Outputs:\n",
    "    -------\n",
    "    FrequencyArray(list): The frequency array of the given k-mer.\n",
    "    \"\"\"\n",
    "\n",
    "    # Define the frequency array of zeros with predetermined length\n",
    "    FrequencyArray = []\n",
    "    for x in range(4**k):\n",
    "        FrequencyArray.append(0)\n",
    "        \n",
    "    # Iterate over the length of the genome to fill the frequency array\n",
    "    for i in range(len(Text) - (k-1)):\n",
    "\n",
    "        text_pat = Text[i:i+k]\n",
    "        num = PatternToNumber(text_pat)\n",
    "        FrequencyArray[num] = FrequencyArray[num] + 1\n",
    "\n",
    "    return FrequencyArray\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "def FindingFrequentWordsBySorting(Text,k, t):\n",
    "    \"\"\"\n",
    "    Searches the sequence Text for the k-mers (sub-sequences of length k)\n",
    "    that occur at least t times in Text.\n",
    "    ------\n",
    "    Inputs: \n",
    "    ------\n",
    "    Text(string): A DNA string for which the frequency array\n",
    "    is computed.\n",
    "    k(int): k-mer length\n",
    "    t(int): minimum number of times that k-mer should occur in Text\n",
    "    ------\n",
    "    Outputs:\n",
    "    -------\n",
    "    FrequentPatterns(list): The list of k-mers that occur at least t times in Text\n",
    "    \"\"\"\n",
    "    \n",
    "    # Initialize the empty list of the Frequency Array \n",
    "    FrequentPatterns = []\n",
    "    Index = [0 for x in range(len(Text)+1 - k)]\n",
    "    Count = [0 for x in range(len(Text)+1 - k)]\n",
    "    \n",
    "    # Loop over the length of the genome sequence (sans last k-mer length)\n",
    "    for i in range(len(Text)+1 - k):\n",
    "        # Initialize a k-mer\n",
    "        Pattern = Text[i:i+k]\n",
    "        # Convert the Pattern into its position in the frequency array\n",
    "        Index[i] = PatternToNumber(Pattern)\n",
    "        Count[i] = 1\n",
    "\n",
    "    # Sort the indexes of the kmers in lexicographic order\n",
    "    Index.sort()\n",
    "    # Loop over the length of the genome sequence (sans last k-mer length)\n",
    "    for i in range(len(Text)+1 - k):\n",
    "        # If sorted index of a given k-mer is next to the identical k-mer,\n",
    "        # increment the counter\n",
    "        if Index[i] == Index[i-1]:\n",
    "            Count[i] = Count[i-1] + 1\n",
    "\n",
    "        if Count[i] >= t:\n",
    "            Pattern = NumberToPattern(Index[i],k)\n",
    "            FrequentPatterns.append(Pattern)\n",
    "        \n",
    "    return list(set(FrequentPatterns))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def BetterClumpFinding(genome, k, L, t):\n",
    "    \"\"\"\n",
    "    Finds patterns forming clumps in a string in a faster way.\n",
    "    The \"clump\" is defined as the k-mer that occurs in the genome segment\n",
    "    of length L at least t-times.\n",
    "    ------\n",
    "    Inputs: \n",
    "    ------\n",
    "    - genome(string): A nucleotide sequence in which\n",
    "    clumps are to be found\n",
    "    - k(int): length of k-mer forming (L,t) clumps in genome\n",
    "    - Genome(string): A nucleotide sequence within which\n",
    "    the occurences of the patern should be searched for\n",
    "    - t(int): Number of times that clump occurs within the window\n",
    "    ------\n",
    "    Outputs:\n",
    "    -------\n",
    "    kmer_list (list): All distinct k-mers forming (L,t)-clumps in Genome\n",
    "    \"\"\"\n",
    "\n",
    "    # Get the length of the genome\n",
    "    gL = len(genome)\n",
    "\n",
    "    # Define empty frequency array\n",
    "    FrequencyArray = list()\n",
    "\n",
    "    # Define empty array\n",
    "    FrequentPatterns = list()\n",
    "\n",
    "    # Loop over the length of the list to initialize it to zeros\n",
    "    Clump = [0 for x in range(4**k)]\n",
    "\n",
    "    # Get the initial window of genome\n",
    "    Text = str(genome[0:L])\n",
    "    # Compute the frequency array\n",
    "\n",
    "    FrequencyArr = ComputingFrequencies(Text, k)\n",
    "\n",
    "    ## Looping\n",
    "    # Loop over the frequency array with the kmers of given length\n",
    "    for i in range(4**k):\n",
    "        if FrequencyArr[i] >= t:\n",
    "            Clump[i] = 1\n",
    "\n",
    "            \n",
    "    for i in range(1, gL+1 - L):\n",
    "        FirstPattern = genome[i-1: i-1+k]\n",
    "        index = PatternToNumber(FirstPattern)\n",
    "        FrequencyArr[index] = FrequencyArr[index] - 1\n",
    "        \n",
    "        LastPattern = genome[i+L-k: i+L]\n",
    "        index = PatternToNumber(LastPattern)\n",
    "        FrequencyArr[index] = FrequencyArr[index] + 1\n",
    "        \n",
    "        if FrequencyArr[index] >= t:\n",
    "            Clump[index] = 1\n",
    "\n",
    "    for i in range(4**k):\n",
    "        if Clump[i] == 1:\n",
    "            Pattern = NumberToPattern(i, k)\n",
    "            FrequentPatterns.append(Pattern)\n",
    "\n",
    "    return FrequentPatterns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "def CheckClumpLength(clump, L, t, k):\n",
    "    \"\"\"\n",
    "    Check if the set of t k-mers fits within the length L.\n",
    "    -------\n",
    "    Inputs:\n",
    "    -------\n",
    "    - clump(string): A sequence within which t k-mers should fit.\n",
    "    - L(int): Length of the sequence within which t k-mers should fit\n",
    "    - t(int): Number of k-mers falling into the sequence clump within length L\n",
    "    - k(int): Length of the k-mer\n",
    "    -------\n",
    "    Outputs:\n",
    "    - (Bool): Logical value\n",
    "    \"\"\"\n",
    "    for i in range(len(clump)-t+1):\n",
    "        if clump[t+i-1] - clump[i] <= L-k:\n",
    "            return True\n",
    "        \n",
    "    return False\n",
    "\n",
    "def FastClumpFinding(genome, k, L, t):\n",
    "    \"\"\"\n",
    "    Finds patterns forming clumps in a string.\n",
    "    The \"clump\" is defined as the k-mer that occurs in the genome segment\n",
    "    of length L at least t times.\n",
    "    ------\n",
    "    Inputs: \n",
    "    ------\n",
    "    - genome(string): A nucleotide sequence in which\n",
    "    clumps are to be found\n",
    "    - k(int): length of k-mer forming (L,t) clumps in genome\n",
    "    - Genome(string): A nucleotide sequence within which\n",
    "    the occurences of the patern should be searched for\n",
    "    - t(int): Number of times that clump occurs within the window\n",
    "    ------\n",
    "    Outputs:\n",
    "    -------\n",
    "    kmer_list (list): All distinct k-mers forming (L,t)-clumps in Genome\n",
    "    \"\"\"\n",
    "\n",
    "    # Define an empty dictionary for storing k-mers\n",
    "    kmer_dict = {}\n",
    "    \n",
    "    # Define an empty list of output clumps\n",
    "    kmers_out = []\n",
    "    \n",
    "    # Iterate over the genome length\n",
    "    for i in range(len(genome)-k+1):\n",
    "        # Test if the given k-mer exists in dict, if yes, increase the count\n",
    "        # for the given k-mer and also store the starting index of the k-mer\n",
    "        # If k-mer is not yet in the dictionary, create a new key for the k-mer and add to count\n",
    "        if genome[i:i+k] in kmer_dict:\n",
    "            \n",
    "            kmer_dict[genome[i:i+k]][0] += 1\n",
    "            kmer_dict[genome[i:i+k]][1].append(i)\n",
    "            \n",
    "        else:\n",
    "            kmer_dict[genome[i:i+k]] = [1, [i]]\n",
    "    # Check the if the candidate k-mers appear at least t-times, \n",
    "    # and check their indicies (where they appear)\n",
    "    kmer_list = [ [kmer[0], kmer[1][1]] for kmer in kmer_dict.items() if kmer[1][0] >= t]\n",
    "    \n",
    "    # Only pick those k-mers of which at least t fall within a sequence of length L    \n",
    "    for kmer_c in kmer_list:\n",
    "        if CheckClumpLength(kmer_c[1], L, t, k):\n",
    "            kmers_out.append(kmer_c[0])\n",
    "    \n",
    "    return sorted(kmers_out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def ClumpFinding(genome, k, L, t):\n",
    "\t\"\"\"\n",
    "\tFinds patterns forming clumps in a string.\t\n",
    "\t------\n",
    "\tInputs: \n",
    "\t------\n",
    "\t- genome(string): A nucleotide sequence in which\n",
    "\tclumps are to be found\n",
    "\t- k(int): length of k-mer forming (L,t) clumps in genome\n",
    "\t- Genome(string): A nucleotide sequence within which\n",
    "\tthe occurences of the patern should be searched for\n",
    "\t- t(int): Number of times that clump occurs within the window\n",
    "\t------\n",
    "\tOutputs:\n",
    "\t-------\n",
    "\tkmer_list (list): All distinct k-mers forming (L,t)-clumps in Genome\n",
    "\t\"\"\"\n",
    "\n",
    "\t# Get the length of the genome\n",
    "\tgL = len(genome)\n",
    "\n",
    "\t# Define empty list of kmers to be outputted\n",
    "\tkmer_list = list()\n",
    "\n",
    "\t# Get the most frequent kmers \n",
    "\tfreq_kmers = list()\n",
    "\n",
    "\t# Iterate over the frequency array\n",
    "\tfor i in range(gL+1 - L):\n",
    "\n",
    "\t\tcount = list()\n",
    "\t\tfreq_kmers = list()\n",
    "\t\t# freq_kmers = list()\n",
    "\t\tText = genome[i:i+L]\n",
    "\t\t# Iterate over the 'Text' to find the patterns of k-mer length\n",
    "\t\tfor j in range(L+1 - k):\n",
    "\t\t\t# Get the pattern of length k\n",
    "\t\t\tPattern = Text[j:j+k]\n",
    "\t\t\t# Count the number of occurences in the string 'Text'\n",
    "\t\t\tcount.append(PatternCount(Text, Pattern))\n",
    "\t\t\tmax_count = max(count)\n",
    "\n",
    "\t\t# Loop over the string to get the patterns of the \n",
    "\t\t# prescribed length with most occurences\n",
    "\t\tfor m in range(L+1 - k):\n",
    "\t\t\tif count[m] == max_count:\n",
    "\t\t\t\tfreq_kmers.append(Text[m:m+k])\n",
    "\n",
    "\t\t# Flatten the list of lists 'count_list'\n",
    "\t\tflat_list = [item for sublist in freq_kmers for item in sublist]\n",
    "\n",
    "\t\t# Use counter to get the dictionary where each kmer\n",
    "\t\t# is a key and number of occurences in the string\n",
    "\t\t# 'genome' is a value\n",
    "\t\tkmer_dict = Counter(freq_kmers)\n",
    "\n",
    "\t\t# Loop over the dictionary of and only select the kmers\n",
    "\t\t# that appear 't' times or more\n",
    "\t\tfor kmer, count in kmer_dict.items():\n",
    "\t\t\tif count >= t:\n",
    "\t\t\t\tkmer_list.append(kmer)\n",
    "\n",
    "\t\t# Remove duplicates from the list\n",
    "\t\tkmer_list= list(set(kmer_list))\n",
    "\t\tkmer_list.sort()\n",
    "\n",
    "\n",
    "\treturn kmer_list"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
